File: lib/engine.ts
import { parse } from 'csv-parse/sync'
import fs from 'fs'
import path from 'path'

export interface Character {
  name: string
  about: string
  image_url: string
  favorites: number
  [key: string]: string | number
}

// Cargar y parsear el CSV
function loadCharacters(): Character[] {
  const csvPath = path.join(process.cwd(), 'data', 'processed_characters.csv')
  const fileContent = fs.readFileSync(csvPath, 'utf-8')
  const records = parse(fileContent, {
    columns: true,
    skip_empty_lines: true,
  })
  
  return records.map((record: Record<string, string>) => ({
    ...record,
    favorites: parseInt(record.favorites) || 0,
  })) as Character[]
}

// Cache de personajes
let charactersCache: Character[] | null = null

export function getCharacters(): Character[] {
  if (!charactersCache) {
    charactersCache = loadCharacters()
  }
  return charactersCache
}

// Obtener columnas de preguntas (excluyendo metadata)
export function getQuestionColumns(): string[] {
  const characters = getCharacters()
  if (characters.length === 0) return []
  
  const excludedColumns = ['name', 'about', 'image_url', 'favorites']
  return Object.keys(characters[0]).filter(col => !excludedColumns.includes(col))
}

// Motor del Akinator
export class AkinatorEngine {
  private currentData: Character[]
  private questions: string[]

  constructor() {
    this.currentData = [...getCharacters()]
    this.questions = getQuestionColumns()
  }

  getBestQuestion(): string | null {
    let bestGain = -1
    let bestQuestion: string | null = null

    for (const q of this.questions) {
      // Si todos los personajes tienen el mismo valor, ignorar
      const values = new Set(this.currentData.map(c => c[q]))
      if (values.size <= 1) continue

      const yesCount = this.currentData.filter(c => c[q] === 1 || c[q] === '1').length
      const noCount = this.currentData.filter(c => c[q] === 0 || c[q] === '0').length

      const score = Math.min(yesCount, noCount) / Math.max(yesCount, noCount + 1e-9)

      if (score > bestGain) {
        bestGain = score
        bestQuestion = q
      }
    }

    return bestQuestion
  }

  updateData(question: string, answer: number): void {
    this.currentData = this.currentData.filter(c => {
      const value = c[question]
      return value === answer || value === String(answer)
    })
  }

  getResult(): Character | null {
    if (this.currentData.length === 0) return null

    if (this.currentData.length > 1) {
      // Ordenar por favoritos y tomar el más popular
      return [...this.currentData].sort((a, b) => b.favorites - a.favorites)[0]
    }

    return this.currentData[0]
  }

  getRemainingCount(): number {
    return this.currentData.length
  }

  toJSON() {
    return {
      currentData: this.currentData,
      questions: this.questions,
    }
  }

  static fromJSON(data: { currentData: Character[]; questions: string[] }): AkinatorEngine {
    const engine = new AkinatorEngine()
    engine.currentData = data.currentData
    engine.questions = data.questions
    return engine
  }
}

File: src/engine.py
import pandas as pd
import numpy as np

class AkinatorEngine:
    def __init__(self, data_path):
        self.df = pd.read_csv(data_path)
        # Solo usamos las columnas de preguntas (desde la 4ta en adelante)
        self.questions = [col for col in self.df.columns if col not in ['name', 'about', 'image_url']]
        self.current_data = self.df.copy()

    def calculate_entropy(self, data):
        """Calcula qué tan incierto es el set actual de personajes."""
        if len(data) <= 1:
            return 0
        # En Akinator, queremos reducir la cantidad de personajes
        # La entropía máxima es cuando hay muchos personajes posibles
        return np.log2(len(data))

    def get_best_question(self):
        best_gain = -1
        best_q = None

        for q in self.questions:
            # SKEW CHECK: Si todos los personajes restantes tienen el mismo valor en esta pregunta,
            # la ganancia es 0. La ignoramos.
            values = self.current_data[q].unique()
            if len(values) <= 1:
                continue

            yes_count = len(self.current_data[self.current_data[q] == 1])
            no_count = len(self.current_data[self.current_data[q] == 0])
            
            score = min(yes_count, no_count) / max(yes_count, no_count + 1e-9)
            
            if score > best_gain:
                best_gain = score
                best_q = q
        
        return best_q

    def update_data(self, question, answer):
        """
        Filtra el dataset según la respuesta del usuario.
        answer: 1 para 'Sí', 0 para 'No'
        """
        self.current_data = self.current_data[self.current_data[question] == answer]

    def get_result(self):
        """Devuelve el personaje si solo queda uno."""
        if len(self.current_data) == 1:
            return self.current_data.iloc[0].to_dict()
        return None